File: Ejercicio4.kt
fun clasificarNotas(notas: Map<String, Int>): Map<String, String> {
    val clasificacion = mutableMapOf<String, String>()

    for ((asignatura, nota) in notas) {
        val clasificacionAsignatura = when (nota) {
            in 95..100 -> "Excelente"
            in 85..94 -> "Notable"
            in 75..84 -> "Bueno"
            in 70..74 -> "Suficiente"
            else -> "Desempeño insuficiente"
        }
        clasificacion[asignatura.toUpperCase()] = clasificacionAsignatura
    }

    return clasificacion
}

fun main() {
    val notasAlumno = mapOf("Lenguajes" to 66, "Redes" to 78, "Programacion Web" to 88, "Inglés" to 95, "Sistemas" to 73)

    val clasificacionNotas = clasificarNotas(notasAlumno)

    println("Notas originales: $notasAlumno")
    println("Clasificación: $clasificacionNotas")
}

File: Ejercicio4.hs
import qualified Data.Map as Map
import Data.Char (toUpper)

-- Función para obtener la calificación correspondiente a una nota
obtenerCalificacion :: Double -> String
obtenerCalificacion nota
  | nota >= 95 = "Excelente"
  | nota >= 85 = "Notable"
  | nota >= 75 = "Bueno"
  | nota >= 70 = "Suficiente"
  | otherwise = "Desempenio insuficiente"

-- Función para procesar el diccionario de asignaturas y notas
procesarCalificaciones :: Map.Map String Double -> Map.Map String String
procesarCalificaciones diccionarioNotas =
  Map.fromList [(map toUpper asignatura, obtenerCalificacion nota) | (asignatura, nota) <- Map.toList diccionarioNotas]

-- Ejemplo de uso
main :: IO ()
main = do
  let notasAlumno = Map.fromList [("Automatas", 100.0), ("Calculo Diferencial", 94.0), ("Ingles", 80.0), ("Programacion Web", 72.0), ("Redes", 30.0)]
  let calificaciones = procesarCalificaciones notasAlumno
  print calificaciones
